src/ORM/DB.php

```diff
@@ -696,4 +712,91 @@ public static function alteration_message($message, $type = "")
     {
         self::get_schema()->alterationMessage($message, $type);
     }
+
+    /**
+     * Check if all tables and field columns for a class exist in the database.
+     *
+     * @param string $class
+     * @return boolean
+     */
+    public static function database_is_ready(string $class): bool
+    {
+        if (!is_subclass_of($class, DataObject::class)) {
+            throw new InvalidArgumentException("$class is not a subclass of " . DataObject::class);
+        }
+
+        // Used for unit tests
+        if (array_key_exists($class, self::$force_database_is_ready)) {
+            return self::$force_database_is_ready[$class];
+        }
+
+        // Don't check again if we already know the db is ready for this class.
+        if (!empty(self::$database_is_ready[$class])) {
+            return true;
+        }
+
+        // Check if all tables and fields required for the class exist in the database.
+        $requiredClasses = ClassInfo::dataClassesFor($class);
+        $schema = DataObject::getSchema();
+        foreach ($requiredClasses as $required) {
+            // Skip test classes, as not all test classes are scaffolded at once
+            if (is_a($required, TestOnly::class, true)) {
+                continue;
+            }
+
+            // if any of the tables aren't created in the database
+            $table = $schema->tableName($required);
+            if (!ClassInfo::hasTable($table)) {
+                return false;
+            }
+
+            // HACK: DataExtensions aren't applied until a class is instantiated for
+            // the first time, so create an instance here.
+            singleton($required);
+
+            // if any of the tables don't have all fields mapped as table columns
+            $dbFields = DB::field_list($table);
+            if (!$dbFields) {
+                return false;
+            }
+
+            $objFields = $schema->databaseFields($required, false);
+            $missingFields = array_diff_key($objFields, $dbFields);
+
+            if ($missingFields) {
+                return false;
+            }
+        }
+        self::$database_is_ready[$class] = true;
```

Comment:
given the class heirachy is iterated over, shouldn't each iterated class be pushed into the `database_is_ready` array so that if I check `MyMember` it will already have a value for the full hierarchy?

In fact, this function looks like a prime candidate to implement recursion for each class in the hierarchy.

Reply:
Yes, that seems reasonable. I think in that case `clear_database_is_ready` will also need to run through a loop of `ClassInfo::dataClassesFor($class)`.